packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+  FullyReleased = "FullyReleased",
+  GracePeriod = "GracePeriod",
+  RecentlyReleased = "RecentlyReleased"
+}
+
+// TODO: make more generic to support non .eth 2nd level domains
+export type Registration = {
+  // Below timestamps are counted in seconds
+  registrationTimestamp: Timestamp | null;
+  expirationTimestamp: Timestamp | null;
+  expiryTimestamp: Timestamp | null; // TODO: Franco, could you please remove this for us?
+
+  primaryStatus: PrimaryRegistrationStatus;
+  secondaryStatus: SecondaryRegistrationStatus | null;
+};
+
+/**
+ * An action that may be taken on a name through a `Registrar`.
+ */
+export const RegistrarAction = {
+  /** Create a new registration. */
+  Register: "register",
+
+  /** Extend an existing registration. */
+  Renew: "renew",
+} as const;
+
+export type RegistrarAction = (typeof RegistrarAction)[keyof typeof RegistrarAction];
+
+/**
+ * The type of registrar charge that is being applied to the domain.
+ */
+export const RegistrarChargeType = {
+  /** The base fee for the name. */
+  BaseFee: "base-fee",
+
+  /** A fee charged for names with special attributes. */
+  SpecialNameFee: "special-name-fee",
+
+  /** A temporary fee charged after a name has been recently released. */
+  TemporaryFee: "recently-released-fee",
+
+  /** A service fee. */
+  ServiceFee: "service-fee",
+} as const;
+
+export type RegistrarChargeType = (typeof RegistrarChargeType)[keyof typeof RegistrarChargeType];
+
+/**
+ * A single distinct charge that must be paid to a `Registrar` to register or
+ * renew a domain.
+ */
+export interface AbstractRegistrarCharge {
+
+  /**
+   * The type of `RegistrarCharge` that is being applied by the `Registrar` to
+   * rent the domain.
+   */
+  type: RegistrarChargeType;
+
+  /**
+   * The price of the `RegistrarCharge` that must be paid to the `Registrar` to
+   * rent the domain.
+   */
+  price: Price;
+
+  /**
+   * The reason why the `RegistrarCharge` is being applied by the `Registrar`
+   * to rent the domain.
+   */
+  reason?: string;
+
+  /**
+   * The period of time that the `RegistrarCharge` is valid for.
+   * 
+   * If `null`, the `RegistrarCharge` is not guaranteed to be time-limited.
+   */
+  validity?: TimePeriod;
+}
+
+export interface RegistrarBaseFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.BaseFee;
+  price: Price;
+  reason?: undefined;
+  validity?: undefined;
+}
+
+export interface RegistrarSpecialNameFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.SpecialNameFee;
+  price: Price;
+  reason: string;
+  validity?: undefined;
+}
+
+export interface RegistrarTemporaryFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.TemporaryFee;
+  price: Price;
+  reason: string;
+  validity: TimePeriod;
+}
+
+export interface RegistrarServiceFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.ServiceFee;
+  price: Price;
+  reason?: string;
+  validity?: undefined;
```

Comment:
```suggestion
  validity: undefined;
```

Reply:
No, that removes the benefit of the approach I wrote has with the type system.